File: csharp/Common/Result.cs
// Delta: Closing the Specification Gap — Chapter 3 / Appendix B
// Add to your shared library project. Use Result<T> for ALL business rule outcomes.
// Never throw exceptions for business rule failures — always return a Result.

namespace YourCompany.Common;

/// <summary>
/// Typed result for operations that return a value.
/// Success carries a value; failure carries a structured error code and message.
/// </summary>
public sealed class Result<T>
{
    public bool IsSuccess { get; }
    public T? Value { get; }
    public string? ErrorCode { get; }
    public string? ErrorMessage { get; }

    private Result(bool ok, T? value, string? code, string? msg)
        => (IsSuccess, Value, ErrorCode, ErrorMessage) = (ok, value, code, msg);

    public static Result<T> Success(T value) =>
        value is null
            ? throw new ArgumentNullException(nameof(value),
                "Success value must not be null - use Result.Success() (non-generic) or Result<Unit> for void operations")
            : new(true, value, null, null);

    public static Result<T> Failure(string errorCode, string errorMessage) =>
        new(false, default, errorCode, errorMessage);

    /// <summary>Transform the success value without altering failure state.</summary>
    public Result<TOut> Map<TOut>(Func<T, TOut> mapper) =>
        IsSuccess
            ? Result<TOut>.Success(mapper(Value!))
            : Result<TOut>.Failure(ErrorCode!, ErrorMessage!);

    /// <summary>Unwrap or throw — only use when failure is a programming error.</summary>
    public T GetValueOrThrow() => IsSuccess ? Value!
        : throw new InvalidOperationException($"[{ErrorCode}] {ErrorMessage}");

    public override string ToString() =>
        IsSuccess ? $"Success({Value})" : $"Failure({ErrorCode}: {ErrorMessage})";
}

/// <summary>
/// Typed result for commands that produce no value (void operations).
/// </summary>
public sealed class Result
{
    public bool IsSuccess { get; }
    public string? ErrorCode { get; }
    public string? ErrorMessage { get; }

    private Result(bool ok, string? code, string? msg)
        => (IsSuccess, ErrorCode, ErrorMessage) = (ok, code, msg);

    public static Result Success() => new(true, null, null);

    public static Result Failure(string errorCode, string errorMessage) =>
        new(false, errorCode, errorMessage);

    public override string ToString() =>
        IsSuccess ? "Success" : $"Failure({ErrorCode}: {ErrorMessage})";
}

/// <summary>Unit type — used as Result&lt;Unit&gt; for void async operations.</summary>
public readonly struct Unit
{
    public static readonly Unit Value = default;
    public static readonly Result<Unit> SuccessResult =
        Result<Unit>.Success(Value);
}

File: typescript/common/result.ts
// Delta: Closing the Specification Gap — Appendix K

export class Result<T> {
  private constructor(
    private readonly _success: boolean,
    private readonly _value?: T,
    readonly errorCode?: string,
    readonly errorMessage?: string,
  ) {}

  static success<T>(value: T): Result<T> {
    if (value === null || value === undefined) {
      throw new Error(
        "success value must not be null/undefined - use Result.voidSuccess() for " +
        "operations that produce no value"
      );
    }
    return new Result(true, value);
  }

  static failure<T>(code: string, message: string): Result<T> {
    return new Result<T>(false, undefined, code, message);
  }

  static voidSuccess(): Result<void> {
    return new Result<void>(true, undefined);
  }

  get isSuccess(): boolean { return this._success; }
  get isFailure(): boolean { return !this._success; }

  get value(): T {
    if (!this._success) {
      throw new Error(`Result is a failure: [${this.errorCode}] ${this.errorMessage}`);
    }
    return this._value!;
  }

  map<U>(fn: (value: T) => U): Result<U> {
    return this._success
      ? Result.success(fn(this._value!))
      : Result.failure(this.errorCode!, this.errorMessage!);
  }

  /**
   * Structural equality. Use this rather than === because Result instances
   * are not reference-equal even when their contents match. C#, Java, and
   * Python implementations get equality from records / dataclasses; in
   * TypeScript it has to be an explicit method.
   */
  equals(other: Result<T>): boolean {
    if (this._success !== other._success) return false;
    if (this._success) return this._value === other._value;
    return this.errorCode === other.errorCode
        && this.errorMessage === other.errorMessage;
  }

  toString(): string {
    return this._success
      ? `Result.success(${this._value})`
      : `Result.failure(${this.errorCode}: ${this.errorMessage})`;
  }
}
